src/async_fuse/memfs/cache/memory_cache.rs

```diff
@@ -0,0 +1,667 @@
+use clippy_utilities::OverflowArithmetic;
+use lockfree_cuckoohash::{pin, LockFreeCuckooHash as HashMap};
+use parking_lot::RwLock;
+use std::collections::HashMap as StdHashMap;
+
+use super::policy::EvictPolicy;
+use super::{Block, BlockCoordinate, IoBlock, Storage};
+use crate::async_fuse::fuse::protocol::INum;
+
+/// Merge the content from `src` to `dst`. This will set `dst` to be dirty.
+fn merge_two_blocks(src: &IoBlock, dst: &mut Block) {
+    let start_offset = src.offset();
+    let end_offset = src.end();
+
+    dst.set_dirty();
+
+    dst.make_mut()
+        .get_mut(start_offset..end_offset)
+        .unwrap_or_else(|| unreachable!("`src` is checked to be in range."))
+        .copy_from_slice(src.as_slice());
+}
+
+/// The in-memory cache, implemented with lockfree hashmaps.
+#[derive(Debug)]
+pub struct InMemoryCache<P, S> {
+    /// The inner map where the cached blocks stored
+    map: HashMap<INum, RwLock<StdHashMap<usize, Block>>>,
+    /// The evict policy
+    policy: P,
+    /// The backend storage
+    backend: S,
+    /// The block size
+    block_size: usize,
+}
+
+impl<P, S> InMemoryCache<P, S> {
+    /// Create a new `InMemoryCache` with specified `policy`, `backend` and `block_size`.
+    pub fn new(policy: P, backend: S, block_size: usize) -> Self {
+        InMemoryCache {
+            map: HashMap::new(),
+            policy,
+            backend,
+            block_size,
+        }
+    }
+
+    /// Get a block from the in-memory cache without fetch it from backend.
+    fn get_block_from_cache(&self, ino: INum, block_id: usize) -> Option<Block> {
+        let guard = pin();
+        let block = self
+            .map
+            .get(&ino, &guard)
+            .and_then(|file_cache| file_cache.read().get(&block_id).cloned());
+        block
+    }
+
+    /// Merge a block into cache in place.
+    /// Return the block if success (the to-be-merged block existing in cache), otherwise returns `None`.
+    fn merge_block_in_place(&self, ino: INum, block_id: usize, src: &IoBlock) -> Option<Block> {
+        let guard = pin();
+
+        let res = if let Some(mut lock) = self.map.get(&ino, &guard).map(RwLock::write) {
+            if let Some(block) = lock.get_mut(&block_id) {
+                merge_two_blocks(src, block);
+                Some(block.clone())
+            } else {
+                None
+            }
+        } else {
+            None
+        };
+        res
+    }
+
+    /// Write a block into the cache without writing through.
+    ///
+    /// If an old block is to be evicted, returns it with its coordinate.
+    fn write_block_into_cache(
+        &self,
+        ino: INum,
+        block_id: usize,
+        block: Block,
+    ) -> Option<(BlockCoordinate, Block)>
+    where
+        P: EvictPolicy<BlockCoordinate>,
+    {
+        let mut empty_file_cache = None;
+        let to_be_evicted = self.policy.put(BlockCoordinate(ino, block_id));
+        let guard = pin();
```

Comment:
I believe there's gap here. Let's assume the cache capacity is 1.
If a thread1 in here put it's own key (1,1) ,and evict the (0,0).
When thread2 in , thread2 put it's own key (2,2), and evict the (1,1), but the thread1 haven't insert (1,1) into map.
So thread2 evict nothing.
Thread1 and thread2 then insert their own key into the map.
So we can left a (1,1) in map but will never be evicted, right ?
What I want to say is , the gap between policy and map is very dangerous.
I tryied to write a test to approve it, but I failed.
Maybe something is wrong?